Share selected currency state across account overview page components

Jinksi opened the pull request:
Resolves #5753
Resolves #5754 

#### Changes proposed in this Pull Request

- The selected currency state is now stored as a URL query param `selected_currency`. This allows the correct currency to remain selected when using the browser "back" button.
- This is updated when the Account Balances currency tab is selected.
- Two new hooks have been added to `clients/overview/hooks.ts`
  - `useSelectedCurrency`: for retrieving and updating the selected currency.
  - `useSelectedCurrencyOverview`: to get the selected currency overview.
- These hooks have been integrated into the Account Balances and Deposit Overview cards.


Screen recording of the state syncing between the cards and after a page refresh:


https://user-images.githubusercontent.com/3147296/231020380-ca9e157e-4ef0-4f62-bd3e-6d3c421e0c39.mov


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Enable the `_wcpay_feature_simplify_deposits_ui` feature flag.
To enable, run `update_option( '_wcpay_feature_simplify_deposits_ui', '1' );` in [WP Console](https://wordpress.org/plugins/wp-console/).
5. Add more than one bank account currency to your store. [See public docs](https://woocommerce.com/document/woocommerce-payments/deposits/deposit-currencies/#adding-multiple-bank-accounts).
6. Enable multi-currencies for more than one of your bank accounts.
7. Purchase some products in each currency, some using the "bypass pending" card number to add funds straight to "available funds": `4000000000000077`.
8. Select each currency in the Account Overview card and ensure the correct currency and amounts are rendered.
9. Viewing the deposit history table will require previous deposits for the selected currency. To achieve this, purchase items with the "available funds" card above, set the deposit schedule to "daily" and wait 1+ days.
10. Refreshing the page should persist the selected currency, once loading has completed.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Jinksi committed:
Add hook to store overview selected currency as WP option

Jinksi committed:
Rename isLoading to specify context

Jinksi committed:
Rename Currency type to CurrencyCode

Jinksi committed:
Add JSDoc to useSelectedCurrency hook

Jinksi committed:
Update selected currency on tab select

Jinksi committed:
Rename onTabSelect for clarity

Jinksi committed:
Add useSelectedCurrencyOverview hook

Jinksi committed:
Use the selected currency for the deposits overview card

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5822` or branch name `add/5753-overview-selected-currency-data-store` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/5753-overview-selected-currency-data-store) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f816b7e7191ff8b51ba88a8b8a22530b2f3a089a
- Build time: 2023-04-11 02:39:03 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Add dummy changelog entry

Jinksi committed:
Update AccountBalances tests with selected currency hook

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Use query param to store selected currency

Jinksi committed:
Remove isLoading from useSelectedCurrency return object

Jinksi committed:
Force TabPanel re-render when selecting currency

to ensure the correct currency tab is selected when selected currency updated.

Jinksi committed:
Add tests for AccountBalances card with currency selection

Jinksi commented:
This PR is mostly ready but on hold, since it will be useful to have #5799 merged for testing purposes.

Jinksi commented on client/components/account-balances/balances-tab-panel.tsx:
```diff
@@ -70,32 +76,52 @@ const AccountBalancesTabPanel: React.FC = () => {
+			<TabPanel
+				tabs={ depositCurrencyTabs }
+				onSelect={ onTabSelect }
+				initialTabName={
+					isSelectedCurrencyValid ? selectedCurrency : undefined
+				}
```
If a string is passed to initialTabName that doesn't match with a tab name, no tab will be selected. So we check to see if it's valid before passing it.

Jinksi commented on client/components/deposits-overview/index.tsx:
```diff
@@ -39,19 +39,12 @@ const DepositsOverviewDetails: React.FunctionComponent< OverviewProps > = ( {
-	} = useAllDepositsOverviews() as AccountOverview.OverviewsResponse;
-
-	const { currencies, account } = overviews;
+	const { account, overview, isLoading } = useSelectedCurrencyOverview();
```
This new hook will return the required logic to return the deposits overview for the selected currency only.

Jinksi commented on client/overview/hooks.ts:
```diff
@@ -0,0 +1,83 @@
+		( currency ) => currency.currency === selectedCurrency
+	);
+
+	const overview = isSelectedCurrencyValid
+		? currencies.find(
+				( currency ) => currency.currency === selectedCurrency
+		  )
+		: currencies[ 0 ];
```
Return the default currency if the selected currency is not in the overviews response.

Jinksi committed:
Don't render balances if currencies are empty

In order to not display inaccurate account balances in the case of an unreachable server.

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Commit:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

mattallan commented:
> This PR is mostly ready but on hold, since it will be useful to have https://github.com/Automattic/woocommerce-payments/pull/5799 merged for testing purposes.

I've just fixed the conflict related to some changes made to that^ PR before it was merged.

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi commented:
I have mentioned in the description that this PR will probably also resolve #5754
We can hold off on this review until #5700 is merged to make sure if preferred.

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Remove unused variable after merge

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Fix missing account variable after merge

Jinksi committed:
Remove redundant type casting

Jinksi committed:
Update snapshot

Jinksi commented on client/components/deposits-overview/test/__snapshots__/index.tsx.snap:
```diff
@@ -122,14 +122,50 @@ exports[`Deposits Overview information Component Renders 1`] = `
-        <span
-          class="wcpay-deposits-overview__heading__description__text"
+        <div
+          class="wcpay-banner-notice wcpay-banner-warning-notice wcpay-deposits-overview__suspended-notice components-notice is-warning"
```
Need to investigate why this snapshot changed.

Jinksi replied:
Fixed in 6b51b10

Jinksi committed:
Rename hook return types for clarity

Jinksi committed:
Fix incorrect rendering of SuspendedDepositNotice

Jinksi committed:
Update tests to use new useSelectedCurrencyOverview hook

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

Jinksi committed:
Remove duplicate lines after merge

Jinksi commented on client/components/account-balances/balances-tab-panel.tsx:
```diff
@@ -74,100 +84,120 @@ const AccountBalancesTabPanel: React.FC = () => {
+					isSelectedCurrencyValid ? selectedCurrency : undefined
+				}
+			>
+				{ ( tab: BalanceTab ) => (
```
This diff looks extensive, however all code within `<TabPanel></TabPanel>` is a copy-paste from `develop`, with indentation changes.

haszari replied:
FYI there's a `Hide whitespace` option in the GitHub diff view, really useful for reviewing changes like this. 

In the gear menu: 

![Screenshot 2023-04-12 at 9 05 22 AM](https://user-images.githubusercontent.com/4167300/231288141-9044f026-a2ee-4a6b-97df-b3ae5a830a83.png)

Jinksi committed:
Show loading state for recent deposit table

Jinksi committed:
Only render recent deposits table if currency exists

Jinksi committed:
Simplify code to improve legibility

Jinksi committed:
Add deposit table row key prop to prevent console error

"Each child in a list should have a unique "key" prop."

Commit:
Merge branch 'develop' into add/5753-overview-selected-currency-data-store

mattallan reviewed (approved):
I've been testing this PR on https://imaginative-hornet.jurassic.ninja/ and it works as expected with multiple and single currency stores.

I tried adding `&selected_currency=trash` to the overview page and confirmed it falls back to the default currency in this case.

The table row loading is a nice touch when changing between currencies and I like the hook approach used for `useSelectedCurrency`. I didn't have any feedback/suggestions on the code and since it's working as expected I'm happy to approve this one.
